Source/Particles/Collision/BinaryCollision/NuclearFusion/NuclearFusionFunc.H

```diff
@@ -182,10 +175,17 @@ public:
             // stride (smaller set size) until we do all collisions (larger set size)
             for (index_type k = coll_idx; k < max_N; k += min_N)
             {
-                // c1k : how many times the current particle of species 1 is paired with a particle
-                // of species 2. Same for c2k.
-                const index_type c1k = (k%NI1 < max_N%NI1) ? c1 + 1: c1;
-                const index_type c2k = (k%NI2 < max_N%NI2) ? c2 + 1: c2;
+
+                // do not check for collision if a particle's weight was
+                // reduced to zero from a previous collision
+                if (idcpu1[ I1[i1] ]==amrex::ParticleIdCpus::Invalid ||
+                    idcpu2[ I2[i2] ]==amrex::ParticleIdCpus::Invalid ) {
+                    p_mask[pair_index] = false;
+                    if (max_N == NI1) { i1 += min_N; }
+                    if (max_N == NI2) { i2 += min_N; }
+                    pair_index += min_N;
+                    continue;
+                }
```

Comment:
Moving the reduction in weight here rather than doing it during the actual collision is very clever! This also gives an easy solution to the main issue I discussed with @RemiLehe about changing the DSMC to use your new collision scheme. I'll make a PR to update DSMC later in the week following the template in this PR.

Reply:
I did have a question though: 
Consider a case where there is 1 low weight particle and 5 high weight particles in a cell. In principle the 1 low weight particle should be considered for a collision with each of the 5 high weight particles (with low collision probability for each), right? So in theory the low weight particle could be scattered multiple times in a single step. However, with the above approach the low weight particle is removed from consideration after the first time it collides, while the new particle created in it's place is not considered for collisions either. Do you not think this discrepancy will lead to problems?

Reply:
Good thought exercise. 

I think the short answer is that one cannot expect correct physics if you only have 1 particle. For example, see the upper right panel of the figure in the PR description. The physics is poorly captured when the number of particles in the deuterium beam approaches zero. A longer answer is given below.

There are two main things to consider. 1) The total probability that the low weight particle has a collision (sum of probability that it collides with each of the 5 particles) should be small. So, the chance that it has two collisions should be really small = small squared, and the chance that it has three collisions should be super small = small cubed, and so forth. 2) The physics is only ever right on average when using MC methods. To get the right physics you would ideally want something like 10-100 similar low weight particles (assuming we are using fusion_multiplier = 1), in which case that part of the distribution function would properly sample the part of the distribution function that the 5 high weight particles represent. 

If you really do just have 1 particle, and you want better statistics, you can always set the fusion_multiplier to some value larger than 1. Additionally, it would be simple to add a flag that would scale the fusion_multiplier locally by Nmax/Nmin, in which case one could resort to something like the original method. This can be useful if one doesn’t care about managing particle number and weights but wants better product species statistics when the number of particles in a cell for one the colliding species is small.

As for not considering collisions of the particles created during the scattering cycle, I think that is also fine. You can just consider it as a strictly explicit treatment. These particles will begin being included in the scattering on the next scattering cycle.

Reply:
Thanks for the explanation! The MCC routine (particles colliding with background fluid) also explicitly assumes only one collision per particle per timestep, so I agree this would be okay under the same conditions.